### PR TITLE
Implements all methods of the ENS registry

### PR DESCRIPTION
**Pull request by nivida**
## Description

This PR does implement all the known methods of the ENS registry and brings more consistency to the ENS module. This as a good base clean up to forward the improvements listed in #3225.

Closed tasks from #3225: 

- [X] Increase code style consistency  
- [X] Support full registry API ( comments #2683 )
- [X] Fix error handling for callbacks (UnhandledPromiseRejectionWarning)

Part of #3225 

### Things I've done

- I've updated the funcDocs to increase the code style consistency
- I've updated the type definitions (also for existing methods because they were misleading)
- I've improved the naming of variables to increase the readability 
- I've added all missing ``Registry`` methods
- I've added the missing but typed ``Resolver`` method ``supportsInterface``
- I've updated the callback signature to be aligned with the rest of the lib but with respecting of backward compatibility
- I've improved the consistency of ``ENS`` module method names and deprecated the legacy ones

### Findings

- ``eth_call`` options can't get passed as expected and are required to for example display the history of owners of a name (You could define the ``defaultBlock`` property to achieve this but this is definitely not an intuitive way to interact with the API).

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## Checklist:

- [x] I have selected the correct base branch.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] Any dependent changes have been merged and published in downstream modules.
- [x] I ran ```npm run dtslint``` with success and extended the tests and types if necessary.
- [x] I ran ```npm run test:unit``` with success and extended the tests if necessary.
- [x] I ran ```npm run build-all``` and tested the resulting file/'s from  ```dist``` folder in a browser.
- [x] I have updated the ``CHANGELOG.md`` file in the root folder.
- [ ] I have tested my code on the live network.

### CODE REVIEW

File: packages/web3-eth-contract/types/index.d.ts
Comment (nivida) [inline packages/web3-eth-contract/types/index.d.ts:24]: I think adding a todo can be done within a PR also if it's out of scope.

File: docs/web3-eth-ens.rst
Comment (nivida) [inline docs/web3-eth-ens.rst:3]: Changed because it is otherwise invalid.

File: docs/web3-eth-ens.rst
Comment (nivida) [inline docs/web3-eth-ens.rst:619]: Optional parameters weren't documented correctly for already implemented methods.

File: docs/web3-eth-ens.rst
Comment (nivida) [inline docs/web3-eth-ens.rst:629]: It's better to forward them to the ``sendTransaction`` documentation instead of duplicating the transaction options documentation here.

File: packages/web3-eth-ens/src/ENS.js
Comment (nivida) [inline packages/web3-eth-ens/src/ENS.js:344]: ``sendOptions`` changed to ``txConfig`` to have more explicit names.

File: packages/web3-eth-ens/src/contracts/Registry.js
Comment (nivida) [inline packages/web3-eth-ens/src/contracts/Registry.js:65]: I'm not sure if this warning is required or if the IDE hint you get is enough.

File: packages/web3-eth-ens/src/contracts/Registry.js
Comment (nivida) [inline packages/web3-eth-ens/src/contracts/Registry.js:124]: It's probably worth to abstract such contract method calls instead of duplicating a lot of code all the time. This could get addressed with a separate PR.

File: packages/web3-eth-ens/src/lib/ResolverMethodHandler.js
Comment (nivida) [inline packages/web3-eth-ens/src/lib/ResolverMethodHandler.js:73]: Required change or we have to use the deprecated and badly named ``resolver`` method

File: test/eth.ens.js
Comment (nivida) [inline test/eth.ens.js:96]: Note to me: Extend test cases/assertion with events.

File: test/eth.ens.js
Comment (nivida) [inline test/eth.ens.js]: Changed to ``async`` function because the ``done`` listener doesn't look clean in this case.

**Review by nivida — COMMENTED** [review]
I have added some explanatory comments and a small note for a required improvement of the ENS "unit" tests.

**Comment by nivida** [comment]
I've kept the ENS e2e tests as they are and will implement them within a separate PR after all the improvements from #3225 are finished.

**Review by cgewecke — COMMENTED** [review]
This looks great. 

There's  a bit of a test gap - coverage is dropping almost a point. Do you think it would be possible to use the fixtures you wrote for #3301 to write some simple tests for this?

I *think* if you put `scripts/e2e.ens.sh` inside the [unit_and_e2e block][1] in ci.sh (maybe right after geth:auto?), it will get picked up by coveralls and we'd have a measurement for these files. 

[EDIT - sorry I just saw this...]

> I've kept the ENS e2e tests as they are and will implement them within a separate PR after all the improvements from #3225 are finished.

Is there a conflict with the other items? It seems normal to add the tests with the work. 

[1]: https://github.com/ethereum/web3.js/blob/1.x/scripts/ci.sh#L21-L27

**Comment by nivida** [comment]
> There's a bit of a test gap - coverage is dropping almost a point. Do you think it would be possible to use the fixtures you wrote for #3301 to write some simple tests for this?
> I think if you put scripts/e2e.ens.sh inside the unit_and_e2e block in ci.sh (maybe right after geth:auto?), it will get picked up by coveralls and we'd have a measurement for these files.

Yep! I will improve the coverage report! Thanks:)

> Is there a conflict with the other items? It seems normal to add the tests with the work.

I've added the base e2e test initially to cover the custom registry case and only added here the unit tests and thought to create all the e2e tests as the last task of #3225. 🤔

**Review by cgewecke — COMMENTED** [review]
I don't completely understand why tests are being deferred for this. Could you explain that a bit more? 

You've made great progress with #3225 - the remaining features seem different from this Registry work. What aspect of their E2E overlaps such that it makes sense to wait until everything's complete to write the tests?

The cov report is showing this:

Screen Shot 2020-01-30 at 1 02 14 AM

**Comment by nivida** [comment]
> I don't completely understand why tests are being deferred for this. Could you explain that a bit more?
> You've made great progress with #3225 - the remaining features seem different from this Registry work. What aspect of their E2E overlaps such that it makes sense to wait until everything's complete to write the tests?

The only reason I had (that’s the reason why I was using the thinking emoticon) is that I thought to do one task after the other from issue #3225 in separate PRs to have them as small as possible. Because we do not have a real unit test in the 1.x and they are a kind of unit/integration tests combined is this also possible without any problem. The ENS e2e tests aren't that important because we already have the ``Contract`` object covered with e2e test cases. Those e2e tests would just help us signal if ENS does change their interface. But as we know does the ENS team a great job and informs the community well about coming interface changes and I don't think they will change the interface in the time between this PR and the e2e PR. :)

In the case it is easier for you to review those e2e tests with having of the latest code changes of the ``web3-eth-ens`` module in the same PR will I definitely add them here. 

> The cov report is showing this:

This is really strange. I've extended the tests and didn't remove any test cases. I will check that closer.

**Comment by nivida** [comment]
@cgewecke 
Oh, I see (https://coveralls.io/builds/28400262/source?filename=packages/web3-eth-ens/src/contracts/Registry.js). The decrease is mostly because of the missing test cases for the error case for those registry methods. I will update those immediately. Thanks! :)

Btw.: It is probably worth to extend our test scripts to print the coverage reports in the CLI of your dev env.

**Comment by cgewecke** [comment]
> Btw.: It is probably worth to extend our test scripts to print the coverage reports in the CLI of your dev env.

Yes! That's a great idea. I will open a PR for that.

**Comment by nivida** [comment]
@cgewecke I've extended the test cases

before (1.x):
Bildschirmfoto 2020-01-30 um 14 15 43

after:
Bildschirmfoto 2020-01-30 um 14 17 29

**Comment by nivida** [comment]
@cgewecke 
FYI:
As mentioned in the [blog post of ENS](https://medium.com/the-ethereum-name-service/ens-registry-migration-bug-fix-new-features-64379193a5a) does the newly deployed registry have some new functionalities. I will implement those in this PR as well. 

Edit:
I will create after this PR a real clean up PR for the ENS classes but definitely with a tight scope.

**Comment by nivida** [comment]
@cgewecke Ready for review *ping* 🎊 :)

Edit:
The reason for those followed commits after is a result of doing a review on my own. I will do such a review next time before I ping you.

File: docs/web3-eth-ens.rst
Comment (cgewecke) [inline docs/web3-eth-ens.rst:179]:
```suggestion
Sets the resolver contract address of a name
```

File: docs/web3-eth-ens.rst
Comment (cgewecke) [inline docs/web3-eth-ens.rst:252]:
```suggestion
Sets the owner of the given name.
```

File: docs/web3-eth-ens.rst
Comment (cgewecke) [inline docs/web3-eth-ens.rst:323]:
```suggestion
Sets the caching TTL (time-to-live) of a name.
```

File: docs/web3-eth-ens.rst
Comment (cgewecke) [inline docs/web3-eth-ens.rst]:
```suggestion
        console.log(receipt); // successfully registered web3.ethereum.eth
```

File: docs/web3-eth-ens.rst
Comment (cgewecke) [inline docs/web3-eth-ens.rst]:
```suggestion
        console.log(receipt); // successfully registered web3.ethereum.eth 
```

Is this comment valid for all three places it appears?
+ setSubnodeOwner
+ setRecord
+ setSubnodeRecord

File: docs/web3-eth-ens.rst
Comment (cgewecke) [inline docs/web3-eth-ens.rst]:
```suggestion
        console.log(receipt); // successfully registered web3.ethereum.eth
```

File: docs/web3-eth-ens.rst
Comment (cgewecke) [inline docs/web3-eth-ens.rst:551]: 👍

File: packages/web3-eth-ens/src/contracts/Registry.js
Comment (cgewecke) [inline packages/web3-eth-ens/src/contracts/Registry.js:478]: Returning after the callback, e.g. having either a callback *or* resolving a promise seems right to me. But do you have any thoughts about why the "both" pattern might have been used in the original? 

Michael Bradley [mentioned][1] a while ago that this occurs elsewhere in the code base and causes weird problems for callback users.

[1]: https://github.com/ethereum/web3.js/issues/3283#issue-540589845

File: packages/web3-eth-ens/src/lib/ResolverMethodHandler.js
Comment (cgewecke) [inline packages/web3-eth-ens/src/lib/ResolverMethodHandler.js:130]: Just double-checking but basically you've changed the callback format here to something more standard?

e.g from
```
callback(error|receipt)
```
to
```
callback(error, receipt)
```

**Review by cgewecke — APPROVED** [review]
This looks great, really nice. Thanks so much for taking the the time to write all those tests - that file is huge.

Left a few rewording suggestions in the docs but otherwise LGTM 💯

File: docs/web3-eth-ens.rst
Comment (nivida) [reply docs/web3-eth-ens.rst]: Oh, thanks! I will remove it:)

Comment (nivida) [reply docs/web3-eth-ens.rst]: The inline comment in the example is correct for the [setSubnodeOwner](https://docs.ens.domains/contract-api-reference/ens#set-subdomain-owner) and [setSubnodeRecord](https://docs.ens.domains/contract-api-reference/ens#set-subdomain-record) method.

I've changed the inline comment of ``setSubnodeOwner`` to: 
``// successfully defined the owner of web3.ethereum.eth``

File: packages/web3-eth-ens/src/contracts/Registry.js
Comment (nivida) [reply packages/web3-eth-ens/src/contracts/Registry.js:478]: This shouldn't give us any problem. The passed callback gets called as expected and it is definitely an anti-pattern to use a callback and promise at the same time.

File: packages/web3-eth-ens/src/lib/ResolverMethodHandler.js
Comment (nivida) [reply packages/web3-eth-ens/src/lib/ResolverMethodHandler.js:130]: Yes, I've increased the consistency as described and kept backward compatibility.

See PR desc:
> I've updated the callback signature to be aligned with the rest of the lib but with respecting of backward compatibility

**Comment by nivida** [comment]
@cgewecke I will wait on your answers on the un-resolved comments and will merge it after. Thanks for your review Chris much appreciated! :)

File: packages/web3-eth-ens/src/contracts/Registry.js
Comment (nivida) [reply packages/web3-eth-ens/src/contracts/Registry.js:478]: As addition: 
> Fix error handling for callbacks (UnhandledPromiseRejectionWarning)

This does solve the task described in the PR desc and noticed while writing the tests.

**Comment by cgewecke** [comment]
@nivida Sorry - yes this has my approval already :) I closed the open questions and approval X 2!

**Comment by nivida** [comment]
@cgewecke 
> @nivida Sorry - yes this has my approval already :) I closed the open questions and approval X 2!

Oh sorry.. I thought a second approval will be shown in the PR overview. #shitHappens

**Pull request merged**
